### DISCUSSION TITLE
M:1 Forecasting Advice

### DISCUSSION THREAD

**Discussion by GabrielDeza** [awslabs/gluonts#1436] [Q&A]
Hi, 
I have several questions about the library.
Background: 
For a university research project, I am trying to predict the log returns of the stock of a company given past X historical log returns as well as twitter activity. Specifically my dataset looks like (N,T) where T= 5 years or 1306 days and N is 15, where 1 of those columns is the log returns and the 14 other columns are twitter activity (volume of tweets, positive sentiment etc). I have this dataset for 11 companies so if you concatenated it all, I have a (11,15,1306) master dataset.

What I am interested in doing is forecasting 1 time series (the log returns) in the future, given 15 time series as an input (ie: past log returns + the 14 twitter related series). In least from looking through the GluonTS library, most models are 1:1 or M:M and not M:1 (ie: what I am looking for)

To make my problem work with GluonTS, I have though of doing the following:

1. Make the twitter features co-variates (ie: feat dynamic real) and train DeepAR on 1 target. This way, you are effectively doing M:1 forecasts. However, co-variates are assumed to be known over all time which is not the case for twitter features (ie: you don't know the tweets in the future). Furthermore, (AFAIK) Gluon-TS is meant for a large number of time series (ie: N in the 1000s, not N=1)

2. Use a global multivariate model like DeepVAR or GPVAR that does M:M. In this case, I have no covariates and I am forecasting all N series in the future. The issue that I had in the past with is scaling. Is it okay if the scale of each time series is different? If it's a problem, should I be normalizing my inputs to have the same scale?

3. If my understanding of GluonTS is correct and there are no M:1 models and (1) and (2) do not make sense, Is anyone aware of models/libraries/papers outside of GluonTS that support M:1 forecasting? I really enjoy Gluon-TS and all the models that it has implemented (DeepAR, Deepstate, TFT etc) and would love to use them but they might not fit my problem statement. I have looked through most of the papers that the models are based off of and they all seem to be M:M or 1:1 (M times).

4. The last use case that I can think of is reformulating option (1) or (2) using the master dataset. Ie: instead of training a model for 1 specific company of the 11 I have data for, I make my target all 11 log returns of the companies and make each ones twitter features as covariates or time series. This way, N is no longer equal to 1 and is now larger which makes it more compatible with the use case of gluon-TS.

Lastly, if you have suggestions outside of the 4 options/concerns I mentioned above about general forecasting, please let me know, any advice is greatly appreciated :D

**Comment by pratikgehlott** [comment]
hey @GabrielDeza can u share some reference on how you did the first point that you have mentioned here. Thanks

**Reply by GabrielDeza** [reply]
Sure.
for (1), my target is just my log return (ie: a 1-dimensional (T,) array) and the covariates are my 14 twitter features which is a 2 dimensional (14,T) array that I pass in as feat_dynamic_real. Depending on if I am doing validation or training, the feat_dynamic_real have to be extended my prediction length into the future (ie: its a (14, T +t) array where t = prediction length). 
The rest is pretty standard (training DeepAR, rolling forecast on my training set, rolling forecast over my testing set, plotting and then metrics calculations.

**Reply by pratikgehlott** [reply]
soo if my datasets original shape was for example (1000, 5) and in this, I have 4 columns as my features and 1 column as my target variable, so in gluon what I will be doing is taking the target column and converting it to an array which is, (let's consider I have 700 data points in the training sample,) so the target will be an array of (700,) and my feat_dynamic_real will be of shape (4,700)? is this correct? @GabrielDeza

**Reply by GabrielDeza** [reply]
Yes. However, as I mentioned feat_dynamic_real are assumed to be known if the future. For example, a feat_dynamic_real could be the day of the week (ie:0,1,2,3,4,5,6 for Monday, Tuesday, ..., Sunday, respectively). This is a feat_dynamic_real because you know what that value is in the future. In my case, I do not know the twitter features in the future so it is an incorrect way of doing things. I only did it this way because I can mimic a M:1 forecast by doing 1:1 forecast and AFAIK there are no M:1 models in gluon-TS. This might not be the case for you.

However, hopefully someone more familiar with gluon-TS than me can chime in if it is true that their are no M:1 to models and if using feat_dynamic_real this way is the best way to mimic M:1 forecasting within Gluon-TS

**Reply by pratikgehlott** [reply]
My explanation is 100% correct, right? , not your case, but in general. @GabrielDeza

**Reply by GabrielDeza** [reply]
It depends if your "features" are known in the future or not. If they are, you can make them feat_dynamic_real. If they are not known in the future, this is an incorrect method. If this is confusing, I suggest looking through the issues page of Gluon-TS where people had similiar concerns or read the problem setup in any of the papers that the models in Gluon-TS are based off of (ie: beginning of section 3 of the [DeepAR paper](https://arxiv.org/pdf/1704.04110.pdf).
 As this is unrelated to the intent of my post, I would appreciate if you opened up a new discussion post about this so that others don't assume this question is answered due to the activity on this thread. Thank you :)

**Comment by jonbeckman** [comment]
@GabrielDeza I have worked on this same issue in the past and my conclusion is to use #2 (Use a global multivariate model like DeepVAR or GPVAR that does M:M). I implemented and exhausted all the other options you mentioned, and while they worked, they didn't work properly. Either the performance was trash or you can't actually predict into the future, unless you supply estimated feature variate values which would result in garbage predictions.

I was hesitant to go down the path of #2 as I was concerned about the wasted compute for predicting M time series when I only needed 1 predicted. However, I have found that this is not a big problem and can simply parse out the single time series that I care about. Additionally, using a global model allows it to learn the embeddings between the variates.

Feel free to ping me here or on the side to discuss this further! I came across your post because I was looking into rolling my dataset, which matched your "rolling forecast on my training set, rolling forecast over my testing set". I would love to know more about how you implemented that.

**Reply by GabrielDeza** [reply]
Hi,
So in the end, I tried both #1 (assuming covariates to be known) and #2 (training the global model). There is actually a third option that I have just thought about but never implemented. You can lag your covariates in option #1 such that time is not violated. For example, imagine you are doing a daily forecast on Monday morning and you are forecasting for the next 3 days ( Monday, Tuesday, Wednesday). In #1, you assume you know the features of Monday, Tuesday and Wednesday by Monday morning, which means you know the future. Instead, you can lag your covariates by your prediction length such that the covariate for Monday is actually Friday , covariate for Tuesday is Saturday and covariate for Wednesday is Sunday. Although this does not violate time (as all your inputs/covariates are in the past), I am unsure if it works in implementation.

Regarding #2 (training global models), I found that although my loss was decreasing significantly (45 -> ~1-2), because the target I really cared about was a much harder to learn time series than the other time series, I found that most of the decrease in loss was fitting the easier features. But this might just because my time series of interest is a very hard time series to fit (log difference in price of a stock) so it's good to hear that it was not a huge problem.

Regarding the rolling forecasts, I ended up implementing them for both the univariate and multivariate models. In both cases, the `generate_rolling_forecast` function did not really work (I believe it's a new feature in the library so it's still fully supported I believe). It's been a while since I wrote this code but here are the big problems with generate_rolling_forecast for both #1 and #2 (and code to fix both cases after that):

#1) In the DeepAR case with 1 target and N covariates, `generate_rolling_forecast` will roll your target but will not roll your covariates. The fix is to do that part yourself. Below is a MWE that does this, where `data` is a numpy N by T array (ie: your covariates and `target_values` is a 1 by T array (ie: your target time series).

```
plotting_training_data = ListDataset([{
    FieldName.TARGET: target,
    FieldName.START: start,
    FieldName.FEAT_DYNAMIC_REAL: FDR}
        for (target, start, FDR) in zip(target_values, start_date, data)], freq ="1B")

dataset_rolled = generate_rolling_dataset(
        dataset=plotting_training_data,
        start_time=pd.Timestamp(df_target.index[0], freq="1B"),
        end_time=pd.Timestamp(df_target.index[train_length+prediction_length], freq = "1B"),
        strategy=StepStrategy(prediction_length=prediction_length, step_size = step_size))
        
for i,train_dict in enumerate(dataset_rolled):
    #step 1 create the start
    start_ = [train_dict['start']]
    #step 2, create the target
    interim_target = train_dict['target'].reshape((1,-1))
    #step 3, create the feat dynamic real (remember, dataset rolled starts from the furthest back in the future to the present)
    fdr =  [data[0][:,:train_length + prediction_length - (i*(step_size))]]
 
    train_data = ListDataset([{
            FieldName.TARGET: target,
            FieldName.START: start,
            FieldName.FEAT_DYNAMIC_REAL: FDR}
                for (target, start, FDR) in zip(interim_target, start_, fdr)], freq ="1B")
    mx.random.seed(seed)
    np.random.seed(seed)
    for test_entry, forecast in zip(train_data, predictor.predict(train_data)):
        to_pandas(test_entry).plot(linewidth=1, color='b', zorder=0, figsize=(13, 7))
        if i !=0:
            forecast.plot(color='m', prediction_intervals=prediction_intervals, show_mean=True, zorder=10)
```

If the above worked, you should be able to get a plot like this:
![HAS_DeepAR_Daily_2_testing_prediction_length=5_BS=20_NBpE=200_epochs=1_parameter=mu_dir=1_eps=0 01_max_iter=1](https://user-images.githubusercontent.com/53381889/120577600-8bce7f00-c3f2-11eb-971a-5c81de88fe68.png)

#2) In the multivariate case, it is a bit more involved because you have the `MultivariateGrouper` now. Atm, you cannot pass in a multivariate dataset into the `generate_rolling_forecast`. The solution is to pass all of your time series individually, make them all "rolling" and then put them together. The code is a bit more complicated and this post is getting long so let me know if the rolling multivariate version is helpful and I will be glad to share a code snippet.

**Reply by jonbeckman** [reply]
Ha we are working on very similar projects. Maybe we should sync up to share knowledge and ideas. 

Thank you for clarifying the rolled datasets. Currently, my train dataset contains all time series from 1 to 3 years ago and my test dataset contains the most recent year. However, my goal is to train on rolling windows throughout the entire dataset, with the test dataset doing the same thing (without overlapping of course). Market behavior changes and while really old data is good for generalization, not training on the latest market trends because it is saved for testing makes me think the model is not performing as well as it can. Is this what you solved?

**Reply by vikasmalhotra08** [reply]
@unibeck Hey there! I been reading this thread and I am interested in this part of your reply. ```I was hesitant to go down the path of #2 as I was concerned about the wasted compute for predicting M time series when I only needed 1 predicted. However, I have found that this is not a big problem and can simply parse out the single time series that I care about. Additionally, using a global model allows it to learn the embeddings between the variates.```

Would you mind sharing how you managed to parse out the one series that you cared about. I am trying to do something similar where I want the DeepVAR model to learn the relationship between 1000+ series but I only care about one of them. I am not sure how would I parse out the result once I have a trained model.
